tfx/components/example_gen/utils.py

```diff
@@ -209,84 +213,183 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+                                 ) -> Tuple[Text, Optional[Text]]:
+  """Search for most recently updated properties given by `spec_list`."""
   split_pattern = os.path.join(uri, split.pattern)
-  if split_pattern.count(SPAN_SPEC) != 1:
-    raise ValueError('Only one {SPAN} is allowed in %s' % split_pattern)
 
-  split_glob_pattern = split_pattern.replace(SPAN_SPEC, '*')
+  split_glob_pattern = split_pattern
+  split_regex_pattern = _glob_to_regex(split_pattern)
+
+  split_glob_pattern = split_glob_pattern.replace(SPAN_SPEC, '*')
+  split_regex_pattern = split_regex_pattern.replace(SPAN_SPEC,
+      '(?P<{}>.*)'.format(SPAN_PROPERTY_NAME))
+
+  if is_match_version:
+    split_glob_pattern = split_glob_pattern.replace(VERSION_SPEC, '*')
+    split_regex_pattern = split_regex_pattern.replace(VERSION_SPEC,
+        '(?P<{}>.*)'.format(VERSION_PROPERTY_NAME))
+
   logging.info('Glob pattern for split %s: %s', split.name, split_glob_pattern)
-  split_regex_pattern = _glob_to_regex(split_pattern).replace(SPAN_SPEC, '(.*)')
   logging.info('Regex pattern for split %s: %s', split.name,
                split_regex_pattern)
-  if re.compile(split_regex_pattern).groups != 1:
-    raise ValueError('Regex should have only one group')
+
+  num_groups = re.compile(split_regex_pattern).groups
+  if is_match_version and num_groups != 2:
+    raise ValueError('Span and version regex should have two groups')
+  elif not is_match_version and num_groups != 1:
+    raise ValueError('Span only regex should have one group')
 
   files = tf.io.gfile.glob(split_glob_pattern)
   latest_span = None
+  latest_version = None
+
   for file_path in files:
     result = re.search(split_regex_pattern, file_path)
     if result is None:
       raise ValueError('Glob pattern does not match regex pattern')
+
+    select_span = result.group(SPAN_PROPERTY_NAME)
     try:
-      span = int(result.group(1))
+      span = int(select_span)
     except ValueError:
-      raise ValueError('Cannot not find span number from %s based on %s' %
-                       (file_path, split_regex_pattern))
+      raise ValueError('Cannot find %s number from %s based on %s' %
+                       (SPAN_PROPERTY_NAME, file_path, split_regex_pattern))
+
+    select_version = None
+    if is_match_version:
+      select_version = result.group(VERSION_PROPERTY_NAME)
+      try:
+        version = int(select_version)
+      except ValueError:
+        raise ValueError('Cannot find %s number from %s based on %s' %
+                         (VERSION_PROPERTY_NAME, file_path,
+                          split_regex_pattern))
+
     if latest_span is None or span >= int(latest_span):
       # Uses str instead of int because of zero padding digits.
-      latest_span = result.group(1)
+      latest_span = select_span
+      latest_version = select_version
+    if latest_version is None or version >= int(latest_version):
+      latest_version = select_version
```

Comment:
how about this? I feel the logic is more straight forward:

if latest_span is None or span > int(latest_span)
      latest_span = select_span
      latest_version = select_version
elif span == int(latest_span) and version > int(latest_version):
      latest_version = select_version

and line275~276 seems will update other span's version to latest span, why test doesn't catch this?

Reply:
yep, it should be elif, I will add a specific test case for this scenario.